test/docker-e2e/e2e_pfm_test.go

```diff
@@ -0,0 +1,374 @@
+	s.Require().True(finalBonB.Equal(initialBonB), "chain-b retained funds: before=%s after=%s", initialBonB.String(), finalBonB.String())
+
+	finalABalance := s.getBalance(ctx, s.chainA, walletA.GetFormattedAddress(), baseDenom)
+	s.Require().True(expectedAAfter.Equal(finalABalance), "chain-a balance mismatch: expected %s got %s", expectedAAfter.String(), finalABalance.String())
+}
+
+func (s *PFMTestSuite) TearDownTest() {
+	ctx := context.Background()
+
+	chains := []*cosmos.Chain{s.chainA, s.chainB, s.chainC}
+	for _, chain := range chains {
+		if chain == nil {
+			continue
+		}
+		if err := chain.Remove(ctx); err != nil {
+			s.T().Logf("Error stopping chain %s: %v", chain.GetChainID(), err)
+		}
+	}
+
+	if s.hermes != nil {
+		if err := s.hermes.Remove(ctx); err != nil {
+			s.T().Logf("Error stopping hermes: %v", err)
+		}
+	}
+}
+
+func (s *PFMTestSuite) getAllBalances(ctx context.Context, chain *cosmos.Chain, address string) sdk.Coins {
+	node := chain.GetNode()
+	if node.GrpcConn == nil {
+		s.T().Logf("GRPC connection is nil for chain %s, returning empty balances", chain.GetChainID())
+		return sdk.Coins{}
+	}
+
+	resp, err := banktypes.NewQueryClient(node.GrpcConn).AllBalances(ctx, &banktypes.QueryAllBalancesRequest{Address: address})
+	s.Require().NoError(err, "failed to query balances for %s on %s", address, chain.GetChainID())
+
+	return resp.Balances
+}
+
+// assertOnlyTwoHopDenom ensures no unexpected IBC denom holds funds.
+func (s *PFMTestSuite) assertOnlyTwoHopDenom(ctx context.Context, chain *cosmos.Chain, address, expectedDenom string) {
+	balances := s.getAllBalances(ctx, chain, address)
+
+	for _, coin := range balances {
+		if strings.HasPrefix(coin.Denom, "ibc/") && coin.Denom != expectedDenom && coin.Amount.GT(sdkmath.ZeroInt()) {
+			s.T().Fatalf("unexpected IBC denom on %s: %s amount=%s", chain.GetChainID(), coin.Denom, coin.Amount.String())
+		}
+	}
+}
+
+// setupPFMInfrastructure sets up three chains and IBC routing required for PFM.
+func (s *PFMTestSuite) setupPFMInfrastructure(ctx context.Context) {
+	t := s.T()
+
+	tag, err := dockerchain.GetCelestiaTagStrict()
+	s.Require().NoError(err, "failed to get celestia tag")
+
+	s.chainA = s.buildChain(ctx, "chain-a", tag)
+	s.chainB = s.buildChain(ctx, "chain-b", tag)
+	s.chainC = s.buildChain(ctx, "chain-c", tag)
+
+	s.hermes, err = relayer.NewHermes(ctx, s.client, t.Name(), s.network, 0, s.logger)
+	s.Require().NoError(err, "failed to create hermes")
+	err = s.hermes.Init(ctx, []tastoratypes.Chain{s.chainA, s.chainB, s.chainC})
+	s.Require().NoError(err, "failed to initialize hermes")
+
+	s.Require().NoError(s.hermes.CreateClients(ctx, s.chainA, s.chainB))
+	s.Require().NoError(s.hermes.CreateClients(ctx, s.chainB, s.chainC))
+
+	s.connAToB, s.chAToB = s.establishConnectionAndChannel(ctx, s.chainA, s.chainB)
+	s.connBToC, s.chBToC = s.establishConnectionAndChannel(ctx, s.chainB, s.chainC)
```

Comment:
I think these should can be done in parallel, the handshake takes quite a bit of time so it might be nice to do chainA -> chainB and chainB -> chainC at the same time, even just using a wait group or something.